fill in a required field without warnings

lepistone opened the pull request.

lepistone committed:
fill in a required field without warnings

pedrobaeza commented on purchase_requisition_multicurrency/model/purchase_requisition.py:
```diff
@@ -38,3 +39,35 @@ class PurchaseRequisition(models.Model):
+
+        # make the column required again
+        self._columns['pricelist_id'].required = True
+        super(PurchaseRequisition, self)._auto_init(cr, context=context)
```
Do you need to call twice the super?

lepistone replied:
yes. the idea is to set it up first as non-required, fill in the empty records, then set it up as required for real. Otherwise
1. we have the error
2. the field is not effectively required, while it should.

pedrobaeza replied:
Well, it seems a bit tricky, but if there is no other way...

lepistone replied:
I agree, but I know no other way. Do you?

pedrobaeza replied:
Have you tried a post_init_hook? I haven't checked if required constraint is applied in these cases.

lepistone replied:
Ok @pedrobaeza, I tested:
- in a `pre_init_hook`, the column does not exist yet, so I cannot fill it
- in a `post_init_hook` the initialization has already been done and it already failed (so I can fill up the column, but I still get the error and the column is not required until the next module update).

Basically, we really need to init twice, and we need to do it exactly when odoo is doing the init (not before, not after).

pedrobaeza replied:
You can also set manually the constraint by a regular SQL instruction via `cr.execute`. It's also tricky, but it seems to me clearer.

lepistone replied:
It doesn't look clearer to me, and anyway I prefer to let the ORM do that. There is a commentary telling why I do this.

Now that I think about it, I do not see how to make the manual SQL method work: there is the same too early/too late issue I described above about the hooks.

Anyway, I'd really love to merge this very soon, because it is blocking #111 (that I want to merge because it's blocking another one).

Thanks!

lepistone committed:
use new() for test data

This avoids problems with required fields (like pricelist_id from
purchase_requisition_multicurrency).

pedrobaeza commented:
OK then. You have my :+1:

gurneyalex commented:
:+1: and I'll fast track the PR.

lepistone commented:
thanks a lot :smile_cat:
